js/conversation_controller.js

```diff
@@ -161,26 +161,32 @@
       if (!conversation) {
         return;
       }
-      if (conversation.isPublic()) {
+
+      // Close group leaving
+      if (conversation.isClosedGroup()) {
+        await conversation.leaveGroup();
+      } else if (conversation.isPublic()) {
         const channelAPI = await conversation.getPublicSendData();
         if (channelAPI === null) {
           log.warn(`Could not get API for public conversation ${id}`);
         } else {
           channelAPI.serverAPI.partChannel(channelAPI.channelId);
         }
+      } else if (conversation.isPrivate()) {
+        const deviceIds = await textsecure.storage.protocol.getDeviceIds(id);
+        await Promise.all(
+          deviceIds.map(deviceId => {
+            const address = new libsignal.SignalProtocolAddress(id, deviceId);
+            const sessionCipher = new libsignal.SessionCipher(
+              textsecure.storage.protocol,
+              address
+            );
+            return sessionCipher.deleteAllSessionsForDevice();
+          })
+        );
       }
+
       await conversation.destroyMessages();
-      const deviceIds = await textsecure.storage.protocol.getDeviceIds(id);
-      await Promise.all(
-        deviceIds.map(deviceId => {
-          const address = new libsignal.SignalProtocolAddress(id, deviceId);
-          const sessionCipher = new libsignal.SessionCipher(
-            textsecure.storage.protocol,
-            address
-          );
-          return sessionCipher.deleteAllSessionsForDevice();
```

Comment:
wouldn't we want to remove closed groups (sessions) from all devices?

Reply:
It's a tricky thing in the following scenario:
1. You are friends with `A`
2. You create a closed group with `A` and `B`
3. You setup a session with `B` since you don't have it
4. You leave the closed group

In that scenario, if you went and removed all your session then you wouldn't be able to speak to `A` at all. We could maybe make it that you remove sessions of the members you aren't friends with?

Reply:
Oh, I see. Wasn't expecting for closed groups to share sessions with private chat (thought they'd have their own sessions).

Reply:
> Oh, I see. Wasn't expecting for closed groups to share sessions with private chat (thought they'd have their own sessions).

Ye, since we were in a rush we have done it that way but i want to eventually move away and use a different session